File: bank-scraper/domain/transactions.py
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic.dataclasses import dataclass

from domain.base import BaseData
from domain.dezimal import Dezimal
from domain.financial_entity import FinancialEntity


class ProductType(str, Enum):
    FUND = "FUND"
    STOCK_ETF = "STOCK_ETF"
    FACTORING = "FACTORING"
    REAL_STATE_CF = "REAL_STATE_CF"


class TxType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    DIVIDEND = "DIVIDEND"
    RIGHT_ISSUE = "RIGHT_ISSUE"
    RIGHT_SELL = "RIGHT_SELL"
    SUBSCRIPTION = "SUBSCRIPTION"
    SWAP_FROM = "SWAP_FROM"
    SWAP_TO = "SWAP_TO"

    INVESTMENT = "INVESTMENT"
    MATURITY = "MATURITY"

    INTEREST = "INTEREST"


@dataclass
class BaseTx(BaseData):
    id: UUID
    ref: str
    name: str
    amount: Dezimal
    currency: str
    type: TxType
    date: datetime
    entity: FinancialEntity
    is_real: bool


@dataclass
class BaseInvestmentTx(BaseTx):
    product_type: ProductType


@dataclass
class AccountTx(BaseTx):
    fees: Dezimal
    retentions: Dezimal
    interest_rate: Dezimal
    avg_balance: Dezimal


@dataclass
class StockTx(BaseInvestmentTx):
    net_amount: Dezimal
    isin: str
    shares: Dezimal
    price: Dezimal
    fees: Dezimal
    ticker: Optional[str] = None
    market: Optional[str] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None
    linked_tx: Optional[str] = None


@dataclass
class FundTx(BaseInvestmentTx):
    net_amount: Dezimal
    isin: str
    shares: Dezimal
    price: Dezimal
    market: str
    fees: Dezimal
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None


@dataclass
class FactoringTx(BaseInvestmentTx):
    net_amount: Dezimal
    fees: Dezimal
    retentions: Dezimal
    interests: Dezimal


@dataclass
class RealStateCFTx(BaseInvestmentTx):
    net_amount: Dezimal
    fees: Dezimal
    retentions: Dezimal
    interests: Dezimal


@dataclass
class Transactions:
    investment: Optional[list[BaseInvestmentTx]] = None
    account: Optional[list[AccountTx]] = None

    def __add__(self, other):
        investment = (self.investment or []) + (other.investment or [])
        account = (self.account or []) + (other.account or [])
        return Transactions(investment=investment, account=account)

File: frontend/app/src/types/transactions.ts
import { EquityType, FundType, ProductType } from "./position"
import { DataSource, EntityOrigin } from "."

export enum TxType {
  BUY = "BUY",
  SELL = "SELL",
  DIVIDEND = "DIVIDEND",
  RIGHT_ISSUE = "RIGHT_ISSUE",
  RIGHT_SELL = "RIGHT_SELL",
  SUBSCRIPTION = "SUBSCRIPTION",
  SWAP_FROM = "SWAP_FROM",
  SWAP_TO = "SWAP_TO",

  TRANSFER_IN = "TRANSFER_IN",
  TRANSFER_OUT = "TRANSFER_OUT",
  SWITCH_FROM = "SWITCH_FROM",
  SWITCH_TO = "SWITCH_TO",

  INVESTMENT = "INVESTMENT",
  REPAYMENT = "REPAYMENT",
  INTEREST = "INTEREST",

  FEE = "FEE",
}

// Base transaction interface
export interface BaseTx {
  id: string
  ref: string
  name: string
  amount: number
  currency: string
  type: TxType
  date: string
  entity: {
    id: string
    name: string
    origin: EntityOrigin
  }
  source: DataSource
  product_type: ProductType
}

// Base investment transaction interface
export type BaseInvestmentTx = BaseTx

// Account transaction interface
export interface AccountTx extends BaseTx {
  fees: number
  retentions: number
  interest_rate?: number
  avg_balance?: number
  net_amount?: number
}

// Stock transaction interface
export interface StockTx extends BaseInvestmentTx {
  net_amount: number
  isin?: string
  shares: number
  price: number
  fees: number
  ticker?: string
  market?: string
  retentions?: number
  order_date?: string
  linked_tx?: string
  equity_type?: EquityType
}

// Fund transaction interface
export interface FundTx extends BaseInvestmentTx {
  net_amount: number
  isin: string
  shares: number
  price: number
  market: string
  fees: number
  retentions?: number
  order_date?: string
  fund_type?: FundType
}

export interface FundPortfolioTx extends BaseInvestmentTx {
  fees: number
  portfolio_name: string
  iban?: string
}

// Factoring transaction interface
export interface FactoringTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

// Real state crowdfunding transaction interface
export interface RealEstateCFTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

// Deposit transaction interface
export interface DepositTx extends BaseInvestmentTx {
  net_amount: number
  fees: number
  retentions: number
}

export interface ManualTransactionBasePayload {
  id: string
  ref: string
  name: string
  amount: number
  currency: string
  type: TxType
  date: string
  entity_id: string
  source: DataSource
  product_type: ProductType
}

export interface ManualAccountTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.ACCOUNT
  fees?: number
  retentions?: number
  interest_rate?: number
  avg_balance?: number
}

export interface ManualStockTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.STOCK_ETF
  ticker?: string
  isin?: string
  shares: number
  price: number
  fees?: number
  retentions?: number
  market?: string
  order_date?: string
}

export interface ManualFundTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.FUND
  isin: string
  shares: number
  price: number
  fees?: number
  retentions?: number
  market?: string
  order_date?: string
}

export interface ManualFundPortfolioTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.FUND_PORTFOLIO
  portfolio_name: string
  fees?: number
  iban?: string
}

export interface ManualFactoringTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.FACTORING
  fees?: number
  retentions?: number
}

export interface ManualRealEstateTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.REAL_ESTATE_CF
  fees?: number
  retentions?: number
}

export interface ManualDepositTransactionPayload
  extends ManualTransactionBasePayload {
  product_type: ProductType.DEPOSIT
  fees?: number
  retentions?: number
}

export type ManualTransactionPayload =
  | ManualAccountTransactionPayload
  | ManualStockTransactionPayload
  | ManualFundTransactionPayload
  | ManualFundPortfolioTransactionPayload
  | ManualFactoringTransactionPayload
  | ManualRealEstateTransactionPayload
  | ManualDepositTransactionPayload

// Transactions container
export interface Transactions {
  investment?: BaseInvestmentTx[]
  account?: AccountTx[]
}

type Tx = AccountTx &
  StockTx &
  FundTx &
  FactoringTx &
  RealEstateCFTx &
  DepositTx

// Transaction query result
export interface TransactionsResult {
  transactions: Tx[]
}

// Transaction query request
export interface TransactionQueryRequest {
  page?: number
  limit?: number
  entities?: string[]
  excluded_entities?: string[]
  product_types?: ProductType[]
  from_date?: string
  to_date?: string
  types?: TxType[]
  historic_entry_id?: string
}
